File: graphWorldFiles/scripts/randomWalk.rb
require 'yaml'
require 'matrix'

if ARGV.size != 1
  puts "Usage : ./graph_metrics.rb [yamlFile]"
end


yamlFile = ARGV[0]
node = Integer(ARGV[1])

def readGraphFile(yamlFile)
  openFile = File.open( yamlFile )
  y = YAML::load_documents( openFile ) { |graph|
    paths = Array.new(graph.length-1)
    for i in 0...graph.length-1
      paths[i] = Array.new(graph.length-1)
      for j in 0...graph.length-1
        if i==j
          paths[i][j] = 0
        elsif graph[i].include?(j)
          paths[i][j] = 1
        else paths[i][j] = 0
        end
      end
    end
  return paths
  }
end

def computeTransitionProbabilities(paths)
  for i in 0...paths.length

    sum = 0
    paths[i].each { |a| sum+=a }

    for j in 0...paths.length
      if paths[i][j] != 0
        paths[i][j] = 1.0/sum
      end
    end
  end
  return paths
end
    

def removeRowAndColumn(original_paths, node)
  paths = original_paths.map { |row| row.clone }
  for i in 0...paths.length
    if i == node
      paths.delete_at(i)
    end
  end

  for i in 0...paths.length
    for j in 0...paths.length+1
      if j == node
        paths[i].delete_at(j)
      end
    end
  end
  
  return paths
end


def computeExpectedWalkTimeToNode(paths, node)
  sub_matrix = removeRowAndColumn(paths, node)
  puts "About to invert the matrix"
  d = (Matrix.identity(sub_matrix.length) - Matrix.rows(sub_matrix))
  puts "Subtracted the matrices"
  m = d.inverse
  puts "We have inverted the matrix!"

  #distance = Matrix.zero(paths.length)
  
  row = (0...m.row_size).map do |i|
    sum = 0
    for j in 0...m.row_size
      sum += m[i, j]
    end

    #distance[i, node] = sum
    sum
  end
  
  return row[0...node] + [0] + row[node..-1]
end
  
p paths = readGraphFile(ARGV[0])
p paths = computeTransitionProbabilities(paths)

puts "About to compute Walk Times"

expectedWalkTimes = (0...paths.length).map do |i|
  r = computeExpectedWalkTimeToNode(paths, i)
  p r
  r
end

p eccentricities = expectedWalkTimes.transpose.map { |r| r.max }
p radius = eccentricities.min
p diameter = eccentricities.max

File: graphWorldFiles/scripts/randomWalk.py
import sys
from numpy  import *
import yaml

if len(sys.argv) != 2:
  print("Usage : ./randomWalk.py [yamlFile]")
  print(sys.argv)
  exit(1)

yamlFile = sys.argv[1]

def readGraphFile(yamlFile):
  f = open(yamlFile)
  dataMap = yaml.safe_load(f)
  f.close()
  del dataMap['Corners']
  numNodes = size(dataMap.keys())
  m = matrix(zeros((numNodes, numNodes)))
  for start, edges in dataMap.iteritems():
    for end in edges:
      m[start, end] = 1
  return m

def computeTransitionProbabilities(paths):
  numNodes = paths.shape[0]
  degrees = sum(paths, axis=0)[0][0]
  for i in range(numNodes):
    for j in range(numNodes):
      paths[i, j] = paths[i, j]/degrees[0, i]
  return paths

def removeRowAndColumn(paths, index):
  r = delete(paths,index, 0)
  r = delete(r, index, 1)
  return r

def computeExpectedWalkTimeToNode(paths, node):
  sub_matrix = removeRowAndColumn(paths, node)
  print(node)
  d = (identity(sub_matrix.shape[0], float) - sub_matrix).I
  row_sums = d.sum(axis=1)
  r = list(array(row_sums)[:, 0])
  return r[0:node] + [0] + r[node:]

paths = computeTransitionProbabilities(readGraphFile(yamlFile))

distance_matrix = matrix([ computeExpectedWalkTimeToNode(paths, i) 
			for i in range(0, paths.shape[0]) ])

eccentricities = distance_matrix.max(axis=0)

print(eccentricities)
print(eccentricities.max())
print(eccentricities.min())
